### ISSUE TITLE
dummy_channel() using sockets fails when OpenTelemetry instrumentation is active

### ISSUE THREAD

**Issue by jenstroeger** [d5h-foss/grpc-interceptor#30]
### What versions of the following are you using?

* `python`: 3.10
* `grpc-interceptor`: 0.13.2
* `grpcio`: 1.51.0
* `protobuf`: 3.20.3

### What operating system (Linux, Windows,...) and version?

Mac, Linux

### What did you do?

Using the `dummy_client()` while global OpenTelemetry instrumentation was active.

### What did you expect to see?

Tests succeed.

### What did you see instead?

Tests failed.

### Anything else we should know about your project / environment?

Adding global OpenTelemetry instrumentation as documented [here](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/grpc/grpc.html):
```python
from opentelemetry.instrumentation.grpc import GrpcInstrumentorClient, GrpcInstrumentorServer

GrpcInstrumentorServer().instrument()  # type: ignore[no-untyped-call]
GrpcInstrumentorClient().instrument()  # type: ignore[no-untyped-call]
```
causes this error:
```
  File "/.../tests/test_interceptors.py", line 34, in test_returns_method_result
    assert client.Execute(DummyRequest(input="foo")).output == "foo"
  File "/.../.venv/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py", line 69, in __call__
    return self._interceptor.intercept_unary(
  File "/.../.venv/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_client.py", line 154, in intercept_unary
    return self._intercept(request, metadata, client_info, invoker)
  File "/.../.venv/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_client.py", line 145, in _intercept
    raise exc
  File "/.../.venv/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/_client.py", line 131, in _intercept
    result = invoker(request, metadata)
  File "/.../.venv/lib/python3.10/site-packages/opentelemetry/instrumentation/grpc/grpcext/_interceptor.py", line 59, in invoker
    return self._base_callable(
  File "/.../.venv/lib/python3.10/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/.../.venv/lib/python3.10/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Exception calling application: not enough values to unpack (expected 2, got 1)"
	debug_error_string = "UNKNOWN:Error received from peer unix:/var/folders/2r/yyz1661s4j91vhdkzktg66z80000gn/T/a8ac6951-2094-4bed-8dfc-678465677448.sock {created_time:"2023-03-26T13:05:06.353245+10:00", grpc_status:2, grpc_message:"Exception calling application: not enough values to unpack (expected 2, got 1)"}"
>
```
I suspect that’s because the `dummy_client` on non Windows hosts opens a UNIX socket rather than a port:

https://github.com/d5h-foss/grpc-interceptor/blob/1c2ec428cb7c7796f68e23cb254129b068aeb839/src/grpc_interceptor/testing/dummy_client.py#L227-L232

The related OpenTelemetry code ([src](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/e4d8f10ecd7bcd29c119af0e3ea7c30d4a383f4b/instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py#L254-L256)) stumbles because
```python
context.peer()
```
returns `unix:` and therefore
```
context.peer().split(",")[0].split(":", 1)[1].rsplit(":", 1)
```
returns `['']` which can’t unpack to `ip` and `port`. If I change to using ports on Mac/Linux, the tests work.

It looks to me like OpenTelemetry doesn’t support channels over sockets, which is probably a feature request for them. However, until then is there any chance you’d welcome a PR that adds a `use_sockets` kwarg to `dummy_channel()` to explicitly enable UNIX sockets and otherwise fall back to ports?

**Comment by d5h** [comment]
Hi @jenstroeger,

Good find and thanks for the details! I agree that OpenTelemetry should probably handle this gracefully, but I'm open to changing the code to use an IPv4 socket instead. That's a better test of how people actually use grpc anyway.

Honestly the only reason it's using a unix domain socket now is that it's an easy way to avoid port conflicts. I still think this is a valid point though. E.g., someone may be running a service on port 50051 at the same time they try to run their tests, or maybe they want to parallelize their tests. (I didn't worry about this for Windows because I've never met anyone who actually runs grpc on Windows, so figured it wasn't worth worrying about.)

So instead of adding a kwarg to tell it to use either a UDS or IPv4 address, I'd rather just have it use an IPv4 address all the time. But to do that, it should try to find an unused port so that tests can be parallelized and it isn't making assumptions about which port is available. So it could do something like trying 10 random ports in the valid port range, and only then giving up.

If you want to make a PR, that would be great! I can also probably do this in a few weeks.

**Comment by jenstroeger** [comment]
> I agree that OpenTelemetry should probably handle this gracefully, but I'm open to changing the code to use an IPv4 socket instead. That's a better test of how people actually use grpc anyway.

Shall I open an issue over there as well?

> So it could do something like trying 10 random ports in the valid port range, and only then giving up.

Do you have a particular port range in mind, or should we just pick a bunch in the upper 50000s?
 
> If you want to make a PR, that would be great! I can also probably do this in a few weeks.

Doesn’t look like a big change, so yes, happy to help!

**Comment by jenstroeger** [comment]
@d5h uhm, I just noticed that the package has no support for Py 3.10 or 3.11 — are you planning to add that? I think Python bytecode might have changed which is why I’m getting 
```
      ast3/Python/ast.c:4514:9: error: implicit declaration of function '_PyUnicode_DecodeUnicodeEscape' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
          v = _PyUnicode_DecodeUnicodeEscape(s, len, NULL, &first_invalid_escape);
              ^
      ast3/Python/ast.c:4514:9: note: did you mean 'PyUnicode_DecodeUnicodeEscape'?
```
when I try to run
```
pip install nox==2022.1.7 toml==0.10.2 poetry==1.0.9
nox
```
in a clean venv 🤔

**Comment by d5h** [comment]
> Shall I open an issue over there as well?

It couldn’t hurt. I don’t know how much they’ll prioritize supporting UDS, but in my mind if someone has a valid grpc service, the telemetry interceptor shouldn’t crash. Plus, the way they implement `instrument()` by monkey-patching `grpc.server`, it means you couldn’t have both an inet server and UDS server in the same process. I’m not sure if anybody does that, but it seems like a valid use case. In any case, it would be good to bring it up with them and they can decide how to prioritize it. Thanks!

> Do you have a particular port range in mind, or should we just pick a bunch in the upper 50000s?

There’s no reason to restrict it to 50000+. This is actually kind of an interesting question because it seems different OS may have [different ranges](https://en.wikipedia.org/wiki/Ephemeral_port) they allow (I’m not an expert on port range portability though). I would go with 32768–60999 since that seems to be at least sort of standard for Linux. It’s probably good enough.

**Comment by d5h** [comment]
> @d5h uhm, I just noticed that the package has no support for Py 3.10 or 3.11 — are you planning to add that? I’m seeing some odd compilation failures…
> 
> when I try to run
> 
> ```
> pip install nox==2022.1.7 toml==0.10.2 poetry==1.0.9
> nox
> ```

Yes this package should support Python 3.10+. However I don’t think this error is related to `grpc-interceptor`. It looks like `nox` is trying to compile CPython and failing, probably because whatever C compiler you have on your system is either too old, too new, or just not supported by CPython.

You can try adding 3.10 and 3.11 to `noxfile.py`. Then run something like `nox -s test-3.10` to only run the tests for the version of Python you have installed. That way it shouldn’t try to compile CPython. You don’t need to test every version of Python locally. There’s a GitHub action for that.

**Comment by d5h** [comment]
The fix for this should be released in 0.15.1 now. Please reopen if you find any other issues upon testing.

**Comment by jenstroeger** [comment]
@d5h thank you for the effort, I appreciate that! Confirming that this works now 👍🏼
